hls4ml/model/optimizer/passes/infer_precision.py

```diff
@@ -0,0 +1,290 @@
+from copy import deepcopy
+
+import numpy as np
+
+from hls4ml.model.optimizer import OptimizerPass
+from hls4ml.model.types import FixedPrecisionType, UnspecifiedPrecisionType
+
+
+class InferPrecisionTypes(OptimizerPass):
+    def match(self, node):
+        for layer_type in node.types.values():
+            if isinstance(layer_type.precision, UnspecifiedPrecisionType):
+                return True
+        return False
+
+    def transform(self, model, node):
+        types_to_infer = []
+        for type_name, type_obj in node.types.items():
+            if isinstance(type_obj.precision, UnspecifiedPrecisionType):
+                types_to_infer.append(type_name)
+
+        inferred_types = self._infer_precision(node, types_to_infer)
+        for type_name in types_to_infer:
+            if type_name not in inferred_types:
+                self._infer_default_type(node, type_name)
+
+        return False  # No model graph changes made
+
+    def _infer_precision(self, node, types_to_infer):
+        node_class = node.class_name
+        if node_class in ['Dense']:
```

Comment:
I winder if it's better to use something like `isinstance(node, Dense)` instead of matching to a class name. Matching to a class name doesn't deal with inheritance. For example, I can see the BatchNormalization matching failing for ApplyAlpha matching.

Reply:
LOL, I did it specifically to avoid this since it is usually not what we want. The ApplyAlpha is an example of this.

Reply:
With QONNX ApplyAlpha does need the precision propagated. It worries me if derived classes by default have different behavior. I think if you want different behavior you explicitly should code it. What happens in ApplyAlpha if you don't forbid it? Should QONNX not use ApplyAlpha in this case?